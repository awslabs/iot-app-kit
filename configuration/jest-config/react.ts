export default {
  setupFilesAfterEnv: ['jest-extended/all'],
  maxWorkers: '50%',
  moduleFileExtensions: ['js', 'ts', 'jsx', 'tsx'],
  collectCoverage: true,
  testEnvironment: 'jsdom',
  transform: {
    '.*\\.(tsx?|jsx?)$': [
      '@swc/jest',
      {
        jsc: {
          transform: {
            react: {
              runtime: 'automatic',
            },
          },
        },
      },
    ],
  },
  transformIgnorePatterns: [],
  moduleNameMapper: {
    '\\.(css|scss)$': 'identity-obj-proxy',
  },
};
