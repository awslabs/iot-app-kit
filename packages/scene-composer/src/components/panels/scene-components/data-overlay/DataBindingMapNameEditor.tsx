import React, { useCallback } from 'react';
import { FormField, Input } from '@awsui/components-react';
import { useIntl } from 'react-intl';
import { isEmpty } from 'lodash';

import { IDataOverlayComponentInternal } from '../../../../store';
import { Component } from '../../../../models/SceneModels';

interface IDataBindingMapNameEditorProps {
  bindingName: string;
  index: number;
  valueDataBindings: Component.ValueDataBindingNamedMap[];
  onUpdateCallback: (componentPartial: Partial<IDataOverlayComponentInternal>, replace?: boolean | undefined) => void;
}

// eslint-disable-next-line no-useless-escape
const INVALID_BINDING_NAME_CHARACTERS = /[\$\{\}\.\[\]]/g;

export const DataBindingMapNameEditor: React.FC<IDataBindingMapNameEditorProps> = ({
  bindingName,
  index,
  valueDataBindings,
  onUpdateCallback,
}) => {
  const { formatMessage } = useIntl();

  const bindingNameError = useCallback(
    (bindingName: string) => {
      if (isEmpty(bindingName)) {
        return formatMessage({ defaultMessage: 'Invalid name', description: 'Input error message' });
      }
      if (valueDataBindings.filter((v) => v.bindingName === bindingName).length > 1) {
        return formatMessage({ defaultMessage: 'Duplicate name', description: 'Input error message' });
      }
      if (bindingName.search(INVALID_BINDING_NAME_CHARACTERS) >= 0) {
        return formatMessage({ defaultMessage: 'Invalid character in the name', description: 'Input error message' });
      }
      return null;
    },
    [valueDataBindings, formatMessage],
  );

  const onBindingNameChange = useCallback(
    (e, index) => {
      const newBindings = [...valueDataBindings];
      newBindings[index] = { ...newBindings[index], bindingName: e.detail.value };
      onUpdateCallback({ valueDataBindings: newBindings });
    },
    [valueDataBindings, onUpdateCallback],
  );

  return (
    <FormField
      errorText={bindingNameError(bindingName)}
      label={formatMessage({ defaultMessage: 'Binding Name', description: 'FormField label' })}
    >
      <Input data-test-id='binding-name-input' value={bindingName} onChange={(e) => onBindingNameChange(e, index)} />
    </FormField>
  );
};
