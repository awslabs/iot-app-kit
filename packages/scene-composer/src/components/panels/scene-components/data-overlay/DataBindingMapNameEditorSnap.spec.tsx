import React from 'react';
import { render } from '@testing-library/react';
import { cloneDeep } from 'lodash';

import { DataBindingMapNameEditor } from './DataBindingMapNameEditor';

describe('DataBindingMapNameEditor', () => {
  const valueDataBindings = [
    {
      bindingName: 'binding-1',
      valueDataBinding: {
        dataBindingContext: 'random-1',
      },
    },
    {
      bindingName: 'binding-2',
      valueDataBinding: {
        dataBindingContext: 'random-2',
      },
    },
  ];
  const onUpdateCallbackMock = jest.fn();

  it('should render binding name correctly', async () => {
    const { container } = render(
      <DataBindingMapNameEditor
        bindingName='binding-1'
        index={0}
        valueDataBindings={valueDataBindings}
        onUpdateCallback={onUpdateCallbackMock}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  it('should render with invalid binding name error', async () => {
    const { container } = render(
      <DataBindingMapNameEditor
        bindingName=''
        index={0}
        valueDataBindings={valueDataBindings}
        onUpdateCallback={onUpdateCallbackMock}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  it('should render with duplicate binding name error', async () => {
    const bindings = cloneDeep(valueDataBindings);
    bindings[0].bindingName = valueDataBindings[1].bindingName;
    const { container } = render(
      <DataBindingMapNameEditor
        bindingName={valueDataBindings[1].bindingName}
        index={0}
        valueDataBindings={bindings}
        onUpdateCallback={onUpdateCallbackMock}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  it('should render with invalid character in name error for $', async () => {
    const { container } = render(
      <DataBindingMapNameEditor
        bindingName='aa $ bb'
        index={0}
        valueDataBindings={valueDataBindings}
        onUpdateCallback={onUpdateCallbackMock}
      />,
    );

    expect(container.querySelector('[errortext="Invalid character in the name"]')).not.toBeNull();
    expect(container).toMatchSnapshot();
  });

  it('should render with invalid character in name error for {', async () => {
    const { container } = render(
      <DataBindingMapNameEditor
        bindingName='aa { bb'
        index={0}
        valueDataBindings={valueDataBindings}
        onUpdateCallback={onUpdateCallbackMock}
      />,
    );

    expect(container.querySelector('[errortext="Invalid character in the name"]')).not.toBeNull();
    expect(container).toMatchSnapshot();
  });

  it('should render with invalid character in name error for }', async () => {
    const { container } = render(
      <DataBindingMapNameEditor
        bindingName='aa } bb'
        index={0}
        valueDataBindings={valueDataBindings}
        onUpdateCallback={onUpdateCallbackMock}
      />,
    );

    expect(container.querySelector('[errortext="Invalid character in the name"]')).not.toBeNull();
    expect(container).toMatchSnapshot();
  });

  it('should render with invalid character in name error for .', async () => {
    const { container } = render(
      <DataBindingMapNameEditor
        bindingName='aa . bb'
        index={0}
        valueDataBindings={valueDataBindings}
        onUpdateCallback={onUpdateCallbackMock}
      />,
    );

    expect(container.querySelector('[errortext="Invalid character in the name"]')).not.toBeNull();
    expect(container).toMatchSnapshot();
  });
});
