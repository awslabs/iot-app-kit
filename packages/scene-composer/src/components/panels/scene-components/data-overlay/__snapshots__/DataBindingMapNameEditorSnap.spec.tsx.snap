// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataBindingMapNameEditor should render binding name correctly 1`] = `
<div>
  <div
    data-mocked="FormField"
    errortext="null"
    label="Binding Name"
  >
    <div
      data-mocked="Input"
      data-test-id="binding-name-input"
      value="binding-1"
    />
  </div>
</div>
`;

exports[`DataBindingMapNameEditor should render with duplicate binding name error 1`] = `
<div>
  <div
    data-mocked="FormField"
    errortext="Duplicate name"
    label="Binding Name"
  >
    <div
      data-mocked="Input"
      data-test-id="binding-name-input"
      value="binding-2"
    />
  </div>
</div>
`;

exports[`DataBindingMapNameEditor should render with invalid binding name error 1`] = `
<div>
  <div
    data-mocked="FormField"
    errortext="Invalid name"
    label="Binding Name"
  >
    <div
      data-mocked="Input"
      data-test-id="binding-name-input"
      value=""
    />
  </div>
</div>
`;

exports[`DataBindingMapNameEditor should render with invalid character in name error for $ 1`] = `
<div>
  <div
    data-mocked="FormField"
    errortext="Invalid character in the name"
    label="Binding Name"
  >
    <div
      data-mocked="Input"
      data-test-id="binding-name-input"
      value="aa $ bb"
    />
  </div>
</div>
`;

exports[`DataBindingMapNameEditor should render with invalid character in name error for . 1`] = `
<div>
  <div
    data-mocked="FormField"
    errortext="Invalid character in the name"
    label="Binding Name"
  >
    <div
      data-mocked="Input"
      data-test-id="binding-name-input"
      value="aa . bb"
    />
  </div>
</div>
`;

exports[`DataBindingMapNameEditor should render with invalid character in name error for { 1`] = `
<div>
  <div
    data-mocked="FormField"
    errortext="Invalid character in the name"
    label="Binding Name"
  >
    <div
      data-mocked="Input"
      data-test-id="binding-name-input"
      value="aa { bb"
    />
  </div>
</div>
`;

exports[`DataBindingMapNameEditor should render with invalid character in name error for } 1`] = `
<div>
  <div
    data-mocked="FormField"
    errortext="Invalid character in the name"
    label="Binding Name"
  >
    <div
      data-mocked="Input"
      data-test-id="binding-name-input"
      value="aa } bb"
    />
  </div>
</div>
`;
